Tests/UnitTests/Core/Propagator/CovarianceEngineTests.cpp

```diff
@@ -164,5 +183,324 @@ BOOST_AUTO_TEST_CASE(covariance_engine_test) {
+  transform = AngleAxis3(angle, axis) * transform;
+
+  transform.translation() = planeSurfaceA->transform(gctx).translation();
+
+  auto planeSurfaceB = Surface::makeShared<PlaneSurface>(transform);
+
+  BoundVector parA;
+  // loc 1 must be zero so we're on the intersection of both surfaces.
+  parA << l0, 0, M_PI / 4., M_PI_2 * 0.9, -1 / 1_GeV, 5_ns;
+
+  BoundMatrix covA;
+  covA.setZero();
+  covA.diagonal() << 1, 2, 3, 4, 5, 6;
+
+  auto [parB, covB] =
+      boundToBound(parA, covA, *planeSurfaceA, *planeSurfaceB, bField);
+
+  // now go back
+  auto [parA2, covA2] =
+      boundToBound(parB, covB, *planeSurfaceB, *planeSurfaceA, bField);
+  CHECK_CLOSE_ABS(parA, parA2, 1e-9);
+  // tolerance is a bit higher here
+  CHECK_CLOSE_COVARIANCE(covA, covA2, 1e-6);
+
+  auto prop = makePropagator(bField);
+  BoundMatrix J = directJacobian(prop, parA, *planeSurfaceA, *planeSurfaceB);
+  BoundMatrix covC = J * covA * J.transpose();
+  CHECK_CLOSE_OR_SMALL((covB.template topLeftCorner<2, 2>()),
+                       (covC.template topLeftCorner<2, 2>()), 1e-6, 1e-9);
+  CHECK_CLOSE_OR_SMALL(covB.diagonal(), covC.diagonal(), 1e-6, 1e-9);
+}
+
+BOOST_DATA_TEST_CASE(CovarianceConversionPerigee,
+                     (bFieldDist ^ bFieldDist ^ bFieldDist ^ angleDist ^
+                      angleDist ^ angleDist ^ posDist ^ posDist ^ posDist ^
+                      locDist ^ locDist ^ angleDist ^ angleDist ^ angleDist) ^
+                         bdata::xrange(100),
+                     Bx, By, Bz, Rx, Ry, Rz, gx, gy, gz, l0, l1, pRx, pRy, pRz,
+                     index) {
+  (void)index;
+  const Vector3 bField{Bx, By, Bz};
+
+  auto planeSurfaceA = MAKE_SURFACE();
+
+  BoundVector parA;
+  parA << l0, l1, M_PI / 4., M_PI_2 * 0.9, -1 / 1_GeV, 5_ns;
+
+  BoundMatrix covA;
+  covA.setZero();
+  covA.diagonal() << 1, 2, 3, 4, 5, 6;
+
+  Vector3 global = planeSurfaceA->localToGlobal(gctx, parA.head<2>());
+
+  Transform3 transform;
+  transform.setIdentity();
+  transform.rotate(AngleAxis3(pRx, Vector3::UnitX()));
+  transform.rotate(AngleAxis3(pRy, Vector3::UnitY()));
+  transform.rotate(AngleAxis3(pRz, Vector3::UnitZ()));
+  transform.translation() = global;
+
+  auto perigee = Surface::makeShared<PerigeeSurface>(transform);
+
+  auto [parB, covB] =
+      boundToBound(parA, covA, *planeSurfaceA, *perigee, bField);
+
+  // now go back
+  auto [parA2, covA2] =
+      boundToBound(parB, covB, *perigee, *planeSurfaceA, bField);
+  CHECK_CLOSE_ABS(parA, parA2, 1e-9);
+  CHECK_CLOSE_COVARIANCE(covA, covA2, 1e-9);
+
+  auto prop = makePropagator(bField);
+  BoundMatrix J = directJacobian(prop, parA, *planeSurfaceA, *perigee);
+  BoundMatrix covC = J * covA * J.transpose();
+  CHECK_CLOSE_ABS((covB.template topLeftCorner<2, 2>()),
+                  (covC.template topLeftCorner<2, 2>()), 1e-7);
+  CHECK_CLOSE_ABS(covB.diagonal(), covC.diagonal(), 1e-7);
+  CHECK_CLOSE_ABS(covB, covC, 1e-7);
```

Comment:
Is this line comparing all elements of the covariance?

Reply:
Yes.

Reply:
> Yes. I added the other lines so I could apply different tolerances but now it might not be needed anymore.

Aha ok. Then it is OK to have them if you like